### PR TITLE
feat(rds): enable data api for aurora cluster

### LINKED ISSUE #28574: aws_rds: Add support for the Data API to DatabaseCluster construct
### Describe the feature

As of December 2023, the Data API is now supported for Aurora Serverless v2 and Aurora Provisioned database clusters (PostgreSQL for now) ([Accouncement Blog](https://aws.amazon.com/blogs/database/introducing-the-data-api-for-amazon-aurora-serverless-v2-and-amazon-aurora-provisioned-clusters/)).

### Use Case

I want a straightforward way to enable and use Data API on my Aurora Serverless v2 cluster in the CDK

### Proposed Solution

I think that the [`DatabaseCluster`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_rds.DatabaseCluster.html) construct should get:
- [`enableDataApi`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_rds.ServerlessCluster.html#enabledataapi) construct prop
-  [`grantDataApiAccess(grantee)`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_rds.ServerlessCluster.html#enabledataapi) method

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2

### Environment details (OS name and version, etc.)

n/a

### PR DESCRIPTION
**Pull request by badmintoncryer**
### Issue # (if applicable)

Closes #28574.

### Reason for this change

Data API is supported for not only Aurora Serverless V1 cluster but also Aurora provisioned and Serverless V2 cluster.
However, it is not supported to enable it for provisioned and Serverless V2 cluster.

### Description of changes

Add `enableDataApi` to `DatabaseClusterBaseProps` and implement `grantDataApiAccess()` to `DatabaseClusterBase` class.

### Description of how you validated changes

Add both unit and integ tests

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-rds/README.md [hunk 1]
```diff
@@ -1166,30 +1166,31 @@ new rds.ServerlessClusterFromSnapshot(this, 'Cluster', {
...
-  },
+// Create an Aurora cluster
+const cluster = new rds.DatabaseCluster(this, 'Cluster', {
+  engine: rds.DatabaseClusterEngine.AURORA_MYSQL,
+  vpc,
+  enableDataApi: true, // Optional - will be automatically set if you call grantDataApiAccess()
 });
 cluster.grantDataApiAccess(fn);
+// It is necessary to grant the function access to the secret associated with the cluster for `DatabaseCluster`.
+cluster.secret!.grantRead(fn);
```
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-rds/README.md:1193]: Curious why is this necessary for Aurora cluster but not for the serverless cluster?

File: packages/aws-cdk-lib/aws-rds/lib/cluster.ts [hunk 2]
```diff
@@ -667,6 +696,7 @@ abstract class DatabaseClusterNew extends DatabaseClusterBase {
       associatedRoles: clusterAssociatedRoles.length > 0 ? clusterAssociatedRoles : undefined,
       deletionProtection: defaultDeletionProtection(props.deletionProtection, props.removalPolicy),
       enableIamDatabaseAuthentication: props.iamAuthentication,
+      enableHttpEndpoint: Lazy.any({ produce: () => this.enableDataApi }),
```
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-rds/lib/cluster.ts:699]: Why is this using `lazy.any..` instead of directly `props.enableDataApi`?

**Review by GavinZZ — COMMENTED** [review]
Thanks for the PR! I have two questions to help me better understand the changes.

File: packages/aws-cdk-lib/aws-rds/README.md (hunk 1 above)
Comment (badmintoncryer) [reply packages/aws-cdk-lib/aws-rds/README.md:1193]: @GavinZZ 

In the serverless cluster, `cluster.secret` is defined in the `ServerlessClusterBase` class (see [here](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-rds/lib/serverless-cluster.ts#L341)), and `secret.grantRead()` is invoked within `ServerlessClusterBase.grantDataApiAccess()`.

However, for the Aurora cluster, `cluster.secret` is not defined in `DatabaseClusterBase` class but rather in `DatabaseClusterNew` class (refer to [this link](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-rds/lib/cluster.ts#L526)).

To align these implementations, it's necessary to move the definition of `cluster.secret` from `DatabaseClusterNew` to `DatabaseClusterBase`. I anticipate this change might be significant. Therefore, I plan to open an issue regarding this modification after the current PR is merged.

I can attempt to make this modification within the current PR if you prefer. What are your thoughts on this?

File: packages/aws-cdk-lib/aws-rds/lib/cluster.ts (hunk 2 above)
Comment (badmintoncryer) [reply packages/aws-cdk-lib/aws-rds/lib/cluster.ts:699]: @GavinZZ 

If we don't use `Lazy` in this point, [this test](https://github.com/aws/aws-cdk/pull/29338/files#diff-5faaee1dfd608a160551274db73518aeab8d44bd3351879b34941a052a0135f9R4108) will be failed `!!   Missing key 'EnableHttpEndpoint'`.
This is because `this.enableDataApi` is modified in the `grantDataApiAccess()`.

Initially, I was surprised by this outcome as well. I had thought that `Lazy` was merely for using CloudFormation's built-in functions. If there's anything I've misunderstood, please don't hesitate to correct me.

Comment (GavinZZ) [reply packages/aws-cdk-lib/aws-rds/lib/cluster.ts:699]: Thanks for the explanation. I think it's a valid use case here to use `Lazy`.

File: packages/aws-cdk-lib/aws-rds/README.md (hunk 1 above)
Comment (GavinZZ) [reply packages/aws-cdk-lib/aws-rds/README.md:1193]: Ah I see, it makes sense here to split it into two PRs.

**Review by GavinZZ — APPROVED** [review]
